src/IO/S3/Client.cpp

```diff
@@ -51,11 +51,43 @@ namespace ErrorCodes
 namespace S3
 {
 
-Client::RetryStrategy::RetryStrategy(uint32_t maxRetries_, uint32_t scaleFactor_, uint32_t maxDelayMs_)
+Int64 calculateDelayOnRetries(const Int64 retries, const Int64 scaleFactor)
+{
+    return (1ul << std::min(retries, 31l)) * scaleFactor;
+}
+
+Client::RetryStrategy::RetryStrategy(uint32_t maxRetries_, ContextPtr context, uint32_t scaleFactor_, uint32_t maxDelayMs_)
     : maxRetries(maxRetries_)
     , scaleFactor(scaleFactor_)
     , maxDelayMs(maxDelayMs_)
 {
+    /// Calculating number of retries based on the `max_execution_time` and delays between retries
+    if (context != nullptr)
+    {
+        Int64 execution_time = context->getSettingsRef().max_execution_time.totalMilliseconds() / 2; /// Here we put total time/2 knowing that we will do retries
```

Comment:
Let's avoid using context here (end everywhere when possible). It's much better to pass max_execution_time as unit param.